nomad/deploymentwatcher/deployment_watcher.go

```diff
@@ -312,100 +342,236 @@ func (w *deploymentWatcher) StopWatch() {
-				}
+		return res, fmt.Errorf("failed to determine last evaluation index for job %q: %v", w.j.ID, err)
+	}
 
-				// Description should include that the job is being rolled back to
-				// version N
-				if j != nil {
-					j, desc = w.handleRollbackValidity(j, desc)
-				} else {
-					desc = structs.DeploymentStatusDescriptionNoRollbackTarget(desc)
-				}
-			}
+	deployment := w.getDeployment()
+	for _, alloc := range allocs {
+		dstate, ok := deployment.TaskGroups[alloc.TaskGroup]
+		if !ok {
+			continue
+		}
 
-			// Update the status of the deployment to failed and create an
-			// evaluation.
-			e := w.getEval()
-			u := w.getDeploymentStatusUpdate(structs.DeploymentStatusFailed, desc)
-			if index, err := w.upsertDeploymentStatusUpdate(u, e, j); err != nil {
-				w.logger.Printf("[ERR] nomad.deployment_watcher: failed to update deployment %q status: %v", w.d.ID, err)
-			} else {
-				w.setLatestEval(index)
+		// Nothing to do for this allocation
+		if alloc.DeploymentStatus == nil || alloc.DeploymentStatus.ModifyIndex <= latestEval {
+			continue
+		}
+
+		// We need to create an eval so the job can progress.
+		if alloc.DeploymentStatus.IsHealthy() {
+			res.createEval = true
+		}
+
+		// If the group is using a deadline, we don't have to do anything.
+		if dstate.ProgressDeadline != 0 {
+			continue
+		}
+
+		// Fail on the first bad allocation
+		if alloc.DeploymentStatus.IsUnhealthy() {
+			// Check if the group has autorevert set
+			if dstate.AutoRevert {
+				res.rollback = true
 			}
-		} else if createEval {
-			// Create an eval to push the deployment along
-			w.createEvalBatched(index)
+
+			// Since we have an unhealthy allocation, fail the deployment
+			res.failDeployment = true
+		}
+
+		// All conditions have been hit so we can break
+		if res.createEval && res.failDeployment && res.rollback {
+			break
+		}
+	}
+
+	return res, nil
+}
+
+// shouldRollback returns whether the job should be rolled back to an earlier
+// stable version by examing the allocations in the deployment.
+func (w *deploymentWatcher) shouldRollback() (bool, error) {
+	snap, err := w.state.Snapshot()
+	if err != nil {
+		return false, err
+	}
+
+	d, err := snap.DeploymentByID(nil, w.deploymentID)
+	if err != nil {
+		return false, err
+	}
+
+	for tg, state := range d.TaskGroups {
+		// We have healthy allocs
+		if state.DesiredTotal == state.HealthyAllocs {
+			continue
 		}
+
+		// We don't need to autorevert this group
+		upd := w.j.LookupTaskGroup(tg).Update
+		if upd == nil || !upd.AutoRevert {
+			continue
+		}
+
+		// Unhealthy allocs and we need to autorevert
+		return true, nil
 	}
+
+	return false, nil
+}
+
+// getDeploymentProgressCutoff returns the progress cutoff for the given
+// deployment
+func getDeploymentProgressCutoff(d *structs.Deployment) time.Time {
+	var next time.Time
+	for _, state := range d.TaskGroups {
```

Comment:
Can this list ever be empty, if so this returns a zero time. Wondering if that affects logic downstream

Reply:
The task groups won't be empty but this function can return a zero time. This is handled by the callers